### PR TITLE
feat(editor): support syntax highlighting for JSP (.jsp) files (#9795)

### LINKED ISSUE #9795: [Feature]: Support syntax highlighting and formatting for JSP (.jsp) files
### Problem or use case

I am working on a project that involves JavaServer Pages (`.jsp` files). Currently, Orca does not officially support or properly recognize `.jsp` file formats. As a result, code rendering (syntax highlighting) and basic editor features do not work well when editing JSP files, making it difficult to develop Java web applications seamlessly.

### Proposed solution

- **Syntax Highlighting:** Add syntax highlighting support for `.jsp` files (supporting the combination of HTML, Java snippets, and JSP tags).

### Alternatives or additional context

_No response_

### PR DESCRIPTION
**Pull request by hwantage**
## Summary

- Adds syntax highlighting and language configuration for JavaServer Pages (`.jsp`, `.jspf`) files in Monaco
  Editor (addresses [#9795](https://github.com/stablyai/orca/issues/9795)).
    - Connects Monaco's HTML Language Service (`registerHTMLLanguageService`) for JSP files to enable HTML tag  completions, hovers, document symbols, and folding, while disabling buffer-modifying features (`rename`, `formatting`) to prevent scriptlet corruption.
    - Fixes comparison operator highlighting issue where `<` and `>` inside scriptlets or EL expressions were
  incorrectly paired as bracket match pairs.
    - Refines tokenization for Java scriptlets (`ignoreCase: false` to avoid colouring variables as types in `List
  list`), adds `(?!%>)` lookahead guards to comments/operators to prevent swallowing the `%>` closing tag, and
  switches operator tokens to `keyword.operator` for proper theme compatibility.
    - Implements state-based tokenization for JSTL / custom tags (`<c:if>`, `<c:forEach>`, etc.) and supports
  embedded `${...}` and `#{...}` EL expressions within tag attributes.
    - Restores standard `nextEmbedded: 'javascript'` and `nextEmbedded: 'css'` stack handling for `<script>` and
  `<style>` blocks to ensure correct keyword colors (`function`, `var`, etc.) and bracket matching.
    - Special thanks to @taehwanis for collaborating on the HTML language service integration and tokenizer
  refinements. (#10858)

### What Changed
 1. **Language Detection & Setup**: Registered `.jsp` and `.jspf` file extensions to the `jsp` language ID in  `language-detect.ts` and `monaco-setup.ts` (routing to `htmlWorker`).
2. **HTML Language Service**: Wired `registerHTMLLanguageService('jsp', undefined, jspHtmlModeConfiguration)` to  provide HTML hovers, completions, and folding while keeping buffer-editing features (`rename`, `formatting`)  disabled.
3. **Bracket Matching**: Excluded `<` and `>` from `brackets`, `autoClosingPairs`, and `surroundingPairs` in `register-jsp.ts` to prevent false bracket pair highlighting on comparison operators.
4. **Tokenization & Robustness**:
       - Switched to case-sensitive matching for Java tokens to prevent `List list` from colouring `list` as a type.
       - Added `(?!%>)` lookahead guards to line comments, block comments, and `%` operators so scriptlet closing  tags (`%>`) are never swallowed.
       - Used `keyword.operator` for theme-compatible operator highlighting.
5. **Custom & JSTL Tag Parsing**: Tokenizes `<c:*>` and custom prefix tags with `@tagRest` / `@tagCustom` state transitions, with support for embedded EL (`${...}`, `#{...}`) and scriptlet expressions in attributes.
6. **Embedded Language Support**: Smooth transition for embedded JavaScript and CSS with `@pop` handling on closing tags.
7. **Unit Tests**: Added 17 comprehensive Vitest test cases in `register-jsp.test.ts` checking Monarch rule matches for operators, JSTL tags, EL expressions, comments, edge cases, and language service registration.

## Screenshots / Evidence

| JSP Syntax & JSTL Tags | Script & Embedded JS |
| --- | --- |
| image| image |

## Testing

- [x] Vitest Unit Tests: `npx vitest run src/renderer/src/lib/monaco-languages/register-jsp.test.ts src/renderer/src/lib/language-detect.test.ts` (Files  2 passed (2), Tests  19 passed (19))
- [x] TypeScript Check: `pnpm run tc:web` (0 errors)
- [x] Linter Check: `npx oxlint` on changed files (0 warnings, 0 errors)
- [x] pnpm lint (0 errors from this change)
- [x] pnpm test (Files  3463 passed | 7 skipped(3470), Tests  37046 passed | 60 skipped (37106))
- [x] pnpm build (success)
- [x] Manual Verification (`pnpm dev`): Tested and verified syntax highlighting, JSTL tags, and bracket matching against real-world legacy JSP files.

## AI Review Report

- **Cross-Platform Compatibility**: Uses standard Monaco Editor extension registration. Compatible across macOS, Linux, and Windows.
- **SSH / Remote Compatibility**: Pure renderer-side Monaco tokenization module. No native process or host filesystem assumptions.
- **Performance Risk**: Efficient regex-based Monarch tokenizer registered at boot. Zero memory leaks or extra network fetches.
- **Security & Reliability**: Pure tokenization inside Monaco's sandboxed renderer.

Closes #9795

### CODE REVIEW

**Comment by taehwanis** [comment]
Hi @hwantage — I work in a large JSP codebase daily, so this PR was exactly the feature I needed and I read through it closely to see if I could use it right away. Three things came up while digging in, in case any are useful:

**1. The HTML language service can be attached to a custom language id.** `registerHTMLLanguageService(languageId, options, modeConfiguration)` is exported from the html contribution, and Monaco calls it for `html`, `handlebars` and `razor` — all "HTML plus template syntax", same as JSP. Without it a tokenizer-only registration gives up tag completion, hovers and folding. Two things I'd flag if you try it: the worker switch in `monaco-setup.ts` needs a `case 'jsp'` in the html group (`MonacoEnvironment.getWorker` takes precedence over Monaco's own fallback, so otherwise every call reaches the plain editor worker), and `rename` is worth turning off — it is the only provider that writes to the buffer, and it derives its edits from parsing the file as HTML.

**2. `operator` doesn't resolve in the built-in themes.** `vs-dark` defines `operator.scss`, `operator.sql` and `operator.swift` but no bare `operator`, so line 93's comparison operators tokenize correctly and still render unstyled. `keyword.operator` — which your `elExpression` already uses on line 106 — falls back to `keyword` and shows up. The test asserting `... as operator` passes either way, which is what makes this easy to miss.

**3. `<script>` blocks lose EL and scriptlets.** Once `nextEmbedded: 'javascript'` takes over, `${...}` and `<%= ... %>` inside the block are tokenized as JavaScript. In a legacy JSP codebase I checked (~1,650 files), about a third embed EL inside `<script>`. Monarch can't re-enter host rules mid-embed so this may just be a known limitation, but the share is big enough to be worth documenting.

I've opened #11309 with an implementation that wires the language service. **Please feel free to lift any of it into this PR** — if the maintainers prefer your shape I'll close mine.

**Comment by hwantage** [comment]
@taehwanis Thanks for the detailed review and code proposal! I've merged your HTML language service integration and tokenizer fixes into this PR (#10858) with Co-authored-by credit.

File: src/renderer/src/lib/monaco-languages/register-jsp.ts [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+      [/[;,.]/, 'delimiter'],
+      [/[a-zA-Z_$][\w$]*/, 'identifier'],
+      [/\s+/, 'white']
+    ],
+
+    javaBlockComment: [
+      [/\*\//, 'comment', '@pop'],
+      [/(?:(?!%>)[^*])+/, 'comment'],
+      [/./, 'comment']
+    ],
```
[the remark this answers is not in the document]
Comment (hwantage) [reply src/renderer/src/lib/monaco-languages/register-jsp.ts:89]: Thanks — good catch. %> must terminate a JSP scriptlet before Java parsing, including inside a Java block
  comment.

  Fixed by adding a %> rule in javaBlockComment that uses @popall, returning the tokenizer directly to the JSP root
  state. I also updated string and character-literal rules so they cannot consume %> either, while preserving the
  valid escaped %\> form.

  Added regression coverage for:

  - /* comment %> */ returning to the root state
  - %> inside string and character literals
  - escaped %\> remaining part of the literal

File: src/renderer/src/lib/monaco-languages/register-jsp.ts [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+      [/\/\/(?:(?!%>)[^\n])*/, 'comment'],
+      [/\/\*/, 'comment', '@javaBlockComment'],
+      [/"([^"\\]|\\.)*"/, 'string'],
+      [/'([^'\\]|\\.)*'/, 'string'],
+      [/\b(?:true|false|null)\b/, 'constant'],
+      [
+        /\b(?:if|else|for|while|do|switch|case|default|break|continue|return|try|catch|finally|throw|throws|new|instanceof|import|package|class|interface|enum|extends|implements|public|private|protected|static|final|abstract|synchronized|assert|this|super)\b/,
+        'keyword'
+      ],
+      [/\b(?:int|long|short|byte|float|double|boolean|char|void|String|Object|List|Map)\b/, 'type'],
```
[the remark this answers is not in the document]
Comment (hwantage) [reply src/renderer/src/lib/monaco-languages/register-jsp.ts]: Thanks for the suggestion. I expanded the JSP Java type token list to cover common collection implementations and
  numeric types, including BigDecimal, BigInteger, HashSet, TreeMap, TreeSet, LinkedHashMap, and LinkedHashSet.

  I also extended the regression test to verify that each added type is tokenized as type, while lowercase variable
  names continue to be treated as identifiers.

**Comment by taehwanis** [comment]
@hwantage First, thanks for folding my suggestions into this PR and for the co-author credit — and your August 1 fixes look right to me. The scriptlet terminator inside block comments and string literals was the subtle one, and you got it.

I wanted to tell you about something before doing it rather than after.

While my #11294 and #11322 were going through, I noticed a pattern in how this repo merges outside contributions: small `language-detect.ts` changes get picked up in a day or two (both of mine were one to three lines, full CI, merged in 24-48h). Our two JSP PRs have been open since late July and CI has never been approved on either one. I think a 570-line new grammar is simply the hardest kind of change for a reviewer to start on in a queue this size.

So I closed my #11309 — it duplicated yours anyway — and opened a two-line fallback instead: #12402 maps `.jsp`/`.jspf` onto the built-in `html` id. No grammar, no worker routing.

To be exact about what that does and does not do, I read the rendered token classes in a running build rather than reasoning from the diff:

- markup, `<style>` and `<script>` blocks get colored
- Java inside `<% %>` comes out as a single `mtk1` span — no tokenization at all
- `${...}` EL is likewise `mtk1`; only the braces pick up bracket-pair colorization

Monaco's html tokenizer has no `<%` rule, so that ceiling is structural rather than something to tune. Scriptlet and EL tokenization is exactly what your PR provides and what an `html` mapping can never reach. The fallback body says that explicitly, links here, and states the two entries should become `'jsp'` once this lands.

One thing I want to flag honestly: there is a chance a maintainer sees the fallback merge and treats #9795 as addressed, then closes this as superseded. A batch of PRs was closed that way on August 3. I would rather that not happen, which is why #12402 is written to point at this PR instead of competing with it.

If you would rather I hold it until yours lands, say so and I will close it — no hard feelings at all. And if you would rather own that two-line change under your own name, I will close mine right away and you can open it. It is the easiest merge on the board and I am happy for it to be yours.

**Comment by hwantage** [comment]
Thank you so much for taking the time to share your thoughts and for keeping an eye on this!

I’ve also been closely following the daily updates in Orca and saw your two recent PRs get merged. Since Orca
is such an active project with commits almost every day, I assume the maintainers are prioritizing more critical updates and improvements, which leaves our PR further down the review queue. The fact that JSP is relatively less common in today's ecosystem might be another factor.

Still, even if it takes some time, I really hope this PR doesn't get closed and eventually gets merged. As a developer who works with JSP and Java code on a daily basis, I’m looking forward to seeing full support land in a release as soon as possible.

I completely welcome and support your fallback PR (#12402)—it will serve as a great interim workaround until this full grammar PR gets officially merged.
Thanks again for your thoughtful feedback and contribution!

**Comment by hwantage** [comment]
@coderabbitai 
`scriptBody` and `styleBody` now pop the embed on `${`, `#{` and `<%[=!]?`, tokenize the island through the existing `elExpression` / `jspScriptlet` rules, and resume the embed on the closer. The two island states are shared between script and style: the state name carries the body to return to and the language to resume(`@embedEl.scriptBody.javascript`).

Regression tests run the real `MonarchTokenizer` via `monarch-tokenizer-test-harness` and assert the per-token language sequence for EL and scriptlets in a script body (including resuming JavaScript on the following line) and for EL in a style body.

Known limits, left as-is: the embedded tokenizer restarts from its initial state on resume, so an island inside a JavaScript string or comment leaves the rest of that literal mis-coloured. `<%-- --%>` inside a body is treated as a scriptlet.
